Add support for ReadableStream async iterable prevent cancel

youennf opened the pull request:
#### 5ef993dbb7e08cc7bc08d1e63303404e139cd40a
<pre>
Add support for ReadableStream async iterable prevent cancel
<a href="https://rdar.apple.com/165607588">rdar://165607588</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=303305">https://bugs.webkit.org/show_bug.cgi?id=303305</a>

Reviewed by Chris Dumez.

We add support for converting async iterable parameters in binding generator.
We rebase the binding tests accordingly.

We also update the async iterator handling of end of iteration.
Previously, we would use whether the promise resolves with undefined to detect the end of iteration.
This does not work with ReadableStreams since undefined is a valid value to queue/dequeue.

For that purpose, we do a preliminary step of introducing an IsFinished object that is handling whether the iterator is finished.
We use the IsFinished state to determine when creating the JS iterator tuple whether done is false or true.

Whenever reject is closed, we mark the IsFinished object as finished.
Ditto for now when the resolved value is undefined.
A follow-up patch will change this for ReadableStream.

We update the binding generator test to align with WebIDL requirement (arguments need to all be optional).

Canonical link: <a href="https://commits.webkit.org/303830@main">https://commits.webkit.org/303830@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/08bb8a659f6c87ce9029d24f10a751854be4439c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133750 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6260 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44933 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141325 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f6b47e91-43b9-4d46-bb6b-6864134598eb) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135620 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6785 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6123 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/102315 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4a472a5c-c264-4439-90c9-37c770768c64) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136697 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119907 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83118 "Passed tests") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/027c1249-24df-44f1-a8cb-ddfc10951652) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/4669 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/2281 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113816 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38023 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143973 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5929 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38605 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110691 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6013 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5069 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110881 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28111 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4525 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116162 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59676 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5982 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/34454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5828 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69446 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6074 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5936 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/bindings/js/JSDOMAsyncIterator.h:
```diff
@@ -139,17 +140,31 @@ template<typename JSWrapper, typename IteratorTraits> class JSDOMAsyncIteratorBa
+        static Ref<IsFinished> create() { return adoptRef(*new IsFinished); }
+
+        operator bool() { return m_value; }
+        void setAsFinished() { m_value = true; }
```
We usually use `markAsFinished()` naming for such things I believe, rather than `setAs..`

youennf committed:
Add support for ReadableStream async iterable prevent cancel
rdar://165607588
https://bugs.webkit.org/show_bug.cgi?id=303305

Reviewed by Chris Dumez.

We add support for converting async iterable parameters in binding generator.
We rebase the binding tests accordingly.

We also update the async iterator handling of end of iteration.
Previously, we would use whether the promise resolves with undefined to detect the end of iteration.
This does not work with ReadableStreams since undefined is a valid value to queue/dequeue.

For that purpose, we do a preliminary step of introducing an IsFinished object that is handling whether the iterator is finished.
We use the IsFinished state to determine when creating the JS iterator tuple whether done is false or true.

Whenever reject is closed, we mark the IsFinished object as finished.
Ditto for now when the resolved value is undefined.
A follow-up patch will change this for ReadableStream.

We update the binding generator test to align with WebIDL requirement (arguments need to all be optional).

Canonical link: https://commits.webkit.org/303830@main
